Proposal PR: Remove the management console (🦄unicorns and 🌈rainbows here)

slimsag opened the pull request:
The management console today is the single most significant barrier that site admins encounter when trying to configure their Sourcegraph instance, which is a major blocker for customer onboarding and ARR growth. This (fully-functioning) PR proposes removing it, making the user experience better, reducing technical complexity (-8k LOC) and closing 18 open issues in the process.

From admins not saving their password (#5186), confusing browser TLS warnings prompts (#6070), it being difficult to access (#1934), complexity poking firewall holes properly (#2478), a general lack of maintenance and significant technical complexity for us (#5760, #5350, #4235, #4213, #2759, #2404, #2322, #2095, #2090, #1822, #1731, #1685, #1435, #1432), even slowing down our CI builds (#6770) -- **the management console overall is a massive source of problems for our users and complexity for us from Engineering, Sales, and overall Business standpoint.**

But before we go around suggesting we remove things willy-nilly, why did we add it in the first place and why does it make sense to remove now?

### Management console: a brief history

In November of last year in PR #966 we introduced the management console. At the time, Sourcegraph configuration was provided through a JSON configuration file mounted into the container and this presented challenges for site admins going through our onboarding / trial process:

- Admins needed to not only figure out the right properties to plug in for e.g. their auth provider and code host (which can be quite complex), but also needed to do so in their editor without any autocomplete or helpful tooltips to guide them.
- After making some edits, it required they restart all containers for the changes to take effect.

All in all, this was a tedious process and made configuration very difficult. To resolve this, we made all configuration dynamic so that restarting services was not required. However, some configuration properties such as authentication if misconfigured could lock the site admin out of their instance and prevent them from fixing their configuration after. To resolve this, we took inspiration from other similar tools and created the management console.

Our hope was that the management console and overall configuration UI would eventually become more intuitive, UI-based / beginner-friendly (instead of just a JSON editor), and more. In practice, this work was de-prioritized in favor of other more important work.

Additionally, the management console introduced security concerns which meant we had to lock it down heavily with a self-signed TLS cert on new instances and a strong auto-generated basic auth password in order for it to be sufficiently secure. You can read more about this in my prior proposal [RFC 63 REVIEW: Authless management console access](https://docs.google.com/document/d/1RkOS4EehLtAXhunTazkjCI4yKi5Hc8eRcHZ_v1Fz_QU/edit?ts=5db76618#heading=h.g2cr8neesagy).

**The management console succeeded in lowering the barrier to onboarding at the time we added it, because it meant we could make a majority of the options (site configuration) editable easily at runtime.**

### What has changed?

We can do better.

We understand our product and the configuration options it must provide more intimately than anyone else. When we added it last year, we closely mimicked best-practices in similar software -- and in the meantime, we have grown to make our product more resilient towards configuration. Nearly every single configuration option since has been made dynamic (editable at runtime without requiring a restart), we have paid more close attention to configuration issues, and most important we have learned what the problems are with the configuration of Sourcegraph on its own merits without mimicking other products.

Before starting this change, I wanted to ensure I had a full grasp on the entire situation from a technical point of view. To do this, I compiled a complete list of all existing critical configuration / management console options and determined whether or not they fit the original criteria for what belongs in the management console (things that, if misconfigured, could lead to the admin being locked out). You can find this [document here](https://docs.google.com/document/d/1IMi3UYQCnxXO8LnP14qUB1HpkRlPHf3k7gdGGOYy7tE/edit).

Once this was written, two things became clear to me:

1. **Not a single critical configuration option can prevent the frontend from starting.** This was not true when we added the management console originally and has important implications I will get to later.
2. **Only 5 configuration options could lock an admin out of the instance:**
    - One, the `licenseKey`, I resolved last night in #7181
    - The remaining 4 would need to be set to ridiculously stupid values in order to lock an admin out in practice. _Removing all auth providers_, _setting session expiry to a very low value_, _adding custom HTML which bricks the frontend web UI_, etc.
    - This is not to say these cases are impossible, but that they are **very unlikely in practice and I have not once heard of an admin ever doing this by mistake.**

Given this data, we can shift around our tradeoffs from what they are today:

- 95% of admins onboarding must go through the painful escape hatch (the management console)
- 5% of admins needing an escape hatch have a nice/friendly UI.

To something more amicable to our goals such as:

- 0% of admins onboarding must go through the painful escape hatch.
- 5% of admins needing an escape hatch still have one, and it's MUCH friendlier than the management console in its current form.

### Solution

1. The management console is removed - gone.
2. All configuration options previously found in the management console are automatically migrated into the site configuration.
3. We no longer have both "critical" and "site" configuration: we now have just "site" configuration and one editor / schema.
4. In 95% of cases, admins can make all configuration edits through the regular web UI without any trouble.
5. In 5% of cases where an admin messed up bad and needs an escape hatch, we provide them a command to edit a file in the Docker container using e.g. `vi` or `nano`. This file has your current site configuration contents, and any edits made to it are synchronized with the database. The file is otherwise ephemeral (i.e. DB remains the source of truth for configuration, the file just allows edits as an escape hatch). How this looks:

> ### Editing your site configuration if you cannot access the web UI
>
> If you are having trouble accessing the web UI, you can make edits to your site configuration by editing a file in the container using the following commands:
>
> #### Single-container Docker instances
>
> ```sh
> docker exec -it $CONTAINER -- nano /site-config.json
> ```
>
> #### Kubernetes cluster instances
>
> ```sh
> kubectl exec -it $FRONTEND_POD -- nano /site-config.json
> ```
>
> Perform your edits, type <kbd>ctrl+x</kbd> and <kbd>y</kbd> to save the changes. They will be applied immediately just as if made through the web UI. If you prefer `vi`, simply replace `nano` in the commands above.

### Next steps

100% of what I described above is already implemented in this PR. This could not have been written as an RFC because the solution was inherently tied to the constraints we have in our product -- which could not be uncovered without implementing this.

If this proposal is accepted, we can land this in 3.11 easily.

- [x] Get feedback from Product to confirm they believe this is a good step forward. cc @sqs @christinaforney 
- [x] Get help from @beyang to resolve issues in the regression test suite -- it relied on the management console API for configuration and must now go through the app itself instead.
- [x] Write and update both documentation and in-product documentation/links.
- [x] Write paired changes for deploy-sourcegraph and deploy-sourcegraph-docker to remove the management console.

### Fixes

- Fixes #5186
- Fixes #6070
- Fixes #1934
- Fixes #2478
- Fixes #5760
- Fixes #5350
- Fixes #4235
- Fixes #4213
- Fixes #2759
- Fixes #2404
- Fixes #2322
- Fixes #2095
- Fixes #2090
- Fixes #1822
- Fixes #1731
- Fixes #1685
- Fixes #1435
- Fixes #1432
- Helps #6770

beyang commented:
- [ ] 1 immediate thing that comes to mind: need to update the onboarding widget so that it links to site config, rather than documentation about how to access the mgt console

![image](https://user-images.githubusercontent.com/1646931/70834893-5bcf8e00-1db0-11ea-963b-efa8b41bbda7.png)

beyang commented on enterprise/cmd/frontend/pre-build.sh:
```diff
@@ -14,6 +14,7 @@ parallel_run() {
 echo "--- yarn root"
 # mutex is necessary since frontend and the management-console can
 # run concurrent "yarn" installs
+# TODO: This is no longer needed since the management console was removed.
```
remove if not needed?

slimsag replied:
Prefer to do this in a separate PR with a lot of cleanup

beyang commented on internal/conf/conf.go:
```diff
@@ -162,9 +164,93 @@ func InitConfigurationServerFrontendOnly(source ConfigurationSource) *Server {
 // edit computation.
 var FormatOptions = jsonx.FormatOptions{InsertSpaces: true, TabSize: 2, EOL: "\n"}
+
+var siteConfigEscapeHatchPath = env.Get("SITE_CONFIG_ESCAPE_HATCH_PATH", "/site-config.json", "Path where the site-config.json escape-hatch file will be written.")
```
Does this interact in a sane manner with the `SITE_CONFIG_FILE` env variable?

slimsag replied:
It does. There is no change in how `SITE_CONFIG_FILE` or `CRITICAL_CONFIG_FILE` behave, and this environment variable will remain undocumented. In practice nobody will set this because the default `/site-config.json` is perfect in all deployments except dev environments where writing to the root FS is not possible due to permissions.

beyang commented on internal/conf/conf.go:
```diff
@@ -162,9 +164,93 @@ func InitConfigurationServerFrontendOnly(source ConfigurationSource) *Server {
 	go defaultClient.continuouslyUpdate(nil)
 	close(configurationServerFrontendOnlyInitialized)
+
+	startSiteConfigEscapeHatchWorker(source)
```
Does this mean every `frontend` instance will sync config to its local FS? How does this work in a Kubernetes cluster where there are multiple `frontend` instances?

More broadly, this package could get imported by any service in Sourcegraph, so we should be cautious about what we introduce to the `init` function here. Any way to scope this so that it *only* runs in exactly 1 process per Sourcegraph installation?

slimsag replied:
> More broadly, this package could get imported by any service in Sourcegraph, so we should be cautious about what we introduce to the init function here. Any way to scope this so that it only runs in exactly 1 process per Sourcegraph installation?

You probably just noticed `Init` in the function name, but this is actually in the _InitConfigurationServer**FrontendOnly**_ function which is only invoked once on server startup and only by the frontend, guaranteeing there is exactly one per frontend process and no more.

beyang commented on internal/conf/conf.go:
```diff
@@ -162,9 +164,93 @@ func InitConfigurationServerFrontendOnly(source ConfigurationSource) *Server {
+func startSiteConfigEscapeHatchWorker(c ConfigurationSource) {
+	var (
+		ctx                                        = context.Background()
+		lastKnownFileContents, lastKnownDBContents string
```
Might be cleaner to encapsulate this state in a struct

slimsag replied:
If it's okay with you, I'd like to do this in a follow-up (there are also more pressing things to cleanup here)

beyang commented on internal/db/confdb/confdb.go:
```diff
@@ -5,16 +5,104 @@ import (
 )
 
+// RunMigrations runs configuration DB table migrations.
+func RunMigrations(ctx context.Context) error {
```
- [ ] TODO(@beyang): test this migration runs properly.

beyang commented:
My understanding is that the backup config file is written to a file within the Docker container (the single Docker container in the single node deployment, or each of the frontend Docker containers in the clustered deployment). What happens if the container fails to start? I understand that we don't think config changes will cause the container to fail to start, but containers may fail to start for some other reason.

A counter-proposal that occurred to me (not yet fully fleshed out): what if we provided a CLI one-liner to extract the config from the database and then instructed the site admin to use `SITE_CONFIG_FILE` and `SITE_CONFIG_ALLOW_EDITS=true`?

felixfbecker commented:
This is a great analysis @slimsag and every proposal to remove/replace/change something should look like this! I.e. why was this added in the first place ➡️ do these reasons still apply. 👏🏻

UX should be the deciding factor here, but I would it would certainly remove code maintenance headaches for me.

beyang commented:
@slimsag a note for later: the docs should clearly mention that all this is not necessary if the admin is using `SITE_CONFIG_FILE` (because then the source of config truth is outside the DB).

slimsag commented:
> What migration, if any, is required by existing customers?

I will add documentation before merging this PR which explains the change to site admins and any manual steps that are needed. The migration is already mentioned in the CHANGELOG somewhat effectively: it is automatic and no manual migration is needed.

There is a small caveat to this which is that users of `SITE_CONFIG_FILE` should copy their config from Sourcegraph back into the file so that we can remove this migration code in the future (but nothing bad will happen if they upgrade without doing that, until we remove this migration code down the line). This is mentioned in the CHANGELOG, and will be mentioned in the docs before merging.

> Does this impact @uwedeportivo's work on metrics (or is that completely separate from the management console)?

There is zero relationship with metrics here., I'm not sure what you might be thinking of with this question. It does supersede the work he did in #6696 though.

slimsag commented:
Me and Beyang went over his questions in a call, responding here for transparency sake:

> Does this mean every frontend instance will sync config to its local FS? How does this work in a Kubernetes cluster where there are multiple frontend instances?

The best way to mentally model / reason about the behavior is this: There is no difference between what we do here with the escape hatch file and with what we do in the web UI in terms of handling conflicts.

In the case of the web UI, we sync the DB contents to the web UI, the admin makes an edit, and we sync that edit to the DB.

In the case of the escape-hatch file, we sync the DB contents to the (ephemeral) file on disk, the admin makes an edit, and we sync that edit to the DB.

The file itself only acts as a mechanism for admins to easily make edits without a web UI.

In terms of conflicting edits, we have the same risks as the web UI:  if two admins edit the site config through the web UI at once, one of their changes would be rejected (they would receive an error). In the case of this approach, the error would appear in the frontend container logs whether the conflict is with someone editing the config through a file on another frontend container or through the web UI.

slimsag commented:
> My understanding is that the backup config file is written to a file within the Docker container (the single Docker container in the single node deployment, or each of the frontend Docker containers in the clustered deployment).

This is correct, it is over-written on startup with the current DB contents (which remains the authoritative source of configuration) and is ephemeral.

> What happens if the container fails to start? I understand that we don't think config changes will cause the container to fail to start, but containers may fail to start for some other reason.

There is an implicit assumption that the frontend will always be accessible and not go down due to invalid configuration changes. I have vetted the code and this holds true today for all configuration, even e.g. borking your auth provider config, etc.

In the ultra rare event the frontend would not start at all (e.g. due to a critical bug or regression we introduce), a site admin could load a blank-slate configuration by setting `SITE_CONFIG_FILE` on the container and it would be loaded into the DB on startup.

> A counter-proposal that occurred to me (not yet fully fleshed out): what if we provided a CLI one-liner to extract the config from the database and then instructed the site admin to use `SITE_CONFIG_FILE` and `SITE_CONFIG_ALLOW_EDITS=true`?

The main reason not to rely on `SITE_CONFIG_FILE` alone is because edits become very tedious, and doing it through an editable file on disk is a nicer UX. Instead of using `SITE_CONFIG_FILE` which involves:

1. Providing the right credentials to your psql DB
2. Downloading & running some script to pull the current file from DB
3. Making your edits in your editor
4. Setting `SITE_CONFIG_FILE` on the container, adding a k8s file mount if in k8s
5. Restarting the containers
6. Undoing your SITE_CONFIG_FILE change once it is fixed (if you don't like loading from a file forever).
7. Done!

We can just:

1. Provide a simple command to SSH into the container and open an editor all at once
2. Make your edits and save
3. Done!

slimsag commented:
>  need to update the onboarding widget so that the last item ("Configure SSO") links to site config, rather than documentation about how to access the mgt console. Should be a quick fix.

Yes, there are multiple other places of in-product docs that link to the management console. I will update these before merging and ask for re-review from you.

unknwon commented on cmd/frontend/internal/app/jscontext/jscontext.go:
```diff
@@ -60,10 +60,10 @@ type JSContext struct {
-	LikelyDockerOnMac bool                         `json:"likelyDockerOnMac"`
-	NeedServerRestart bool                         `json:"needServerRestart"`
-	DeployType        string                       `json:"deployType"`
+	Critical          schema.SiteConfiguration `json:"critical"` // public subset of critical configuration
```
1. Is there any concern to expose all of the configuration? (the comment suggests "public subset")
2. Should `Critical` be renamed to `PublicConfig`?

slimsag replied:
1. No, there is no risk, this is only the data type in use. You can see where we construct the actual value in the `publicCriticalConfiguration` function below here (https://github.com/sourcegraph/sourcegraph/pull/7197/files/4be74b77c2ef3367a76c4a6eff5724201a28cda3#diff-507fb3079e6d8fde51150a299cb10c54L184).
2. Yes, it should be renamed / cleaned up. There is much more cleanup that can be done but this PR is quite large so I am hoping to merge it and do this cleanup in a subsequent PR which is less risky.

unknwon commented on cmd/frontend/graphqlbackend/site_alerts.go:
```diff
@@ -62,8 +62,8 @@ func (r *siteResolver) Alerts(ctx context.Context) ([]*Alert, error) {
-		if c.Critical.ExternalURL == "" {
-			problems = append(problems, conf.NewCriticalProblem("`externalURL` is required to be set for many features of Sourcegraph to work correctly."))
+		if c.ExternalURL == "" {
+			problems = append(problems, conf.NewSiteProblem("`externalURL` is required to be set for many features of Sourcegraph to work correctly."))
```
I was going to ask if do we still need [critical type problems](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@487d0d24c0e8ed1c1e8cd8e20ce9c30905dd17dd/-/blob/internal/conf/validate.go#L53), but on a second thought, critical problems are still useful because some configurations if not correct could cause Sourcegraph not working properly. 

Therefore, I would revert back to use `NewCriticalProblem` on this line.

slimsag replied:
We don't need critical type problems anymore, we actually do not need anything with "critical" in the name anymore -- critical configuration is 100% gone after this change and we only have site configuration now.

Some code for "critical" things still remains in this PR because there is _a lot of it_ and I hope to remove this code in a subsequent PR to avoid this one becoming even bigger (which I think is more risky).

> critical problems are still useful because some configurations if not correct could cause Sourcegraph not working properly (wording of site alerts would be changed too, e.g. "some problems in site configuration", "critical problems in site configuration").

Previously we treated these problems as "severe" in any case: if site config is messed up, it can be equally as bad for your Sourcegraph instance as if "critical" config is messed up. I think we always want to display config problems as equal severity because no site admin should ignore config problems (e.g. we should not have "warning" and "error" problem levels, just "problems")

unknwon replied:
Make senses, thanks!

slimsag commented on dev/Procfile:
```diff
@@ -5,7 +5,7 @@ searcher: searcher
 symbols: ./cmd/symbols/build.sh execute
 github-proxy: github-proxy
-frontend: env CONFIGURATION_MODE=server frontend
+frontend: env CONFIGURATION_MODE=server SITE_CONFIG_ESCAPE_HATCH_PATH=$HOME/.sourcegraph/site-config.json frontend
```
No, this file is written on server startup.

slimsag commented on enterprise/cmd/frontend/authz_test.go:
```diff
@@ -134,12 +134,12 @@ func Test_authzProvidersFromConfig(t *testing.T) {
 			},
 			expAuthzAllowAccessByDefault: false,
-			expSeriousProblems:           []string{"Did not find authentication provider matching \"https://gitlab.mine\". Check the [management console](https://docs.sourcegraph.com/admin/management_console) to verify an entry in [`auth.providers`](https://docs.sourcegraph.com/admin/auth) exists for https://gitlab.mine."},
+			expSeriousProblems:           []string{"Did not find authentication provider matching \"https://gitlab.mine\". Check the [site configuration](https://docs.sourcegraph.com/admin/config/site_config) to verify an entry in [`auth.providers`](https://docs.sourcegraph.com/admin/auth) exists for https://gitlab.mine."},
```
Good point!

slimsag replied:
Done

slimsag commented:
> Do we provide or is it necessary to provide downgrade migration if user wants to downgrade from 3.11 to 3.10 for example?

Good question, I will update the migration documentation to reflect this.

The answer: If a user wants to downgrade, they can indeed do so just fine, but after downgrading the critical configuration will be just `{"migrated": true}` and they will need to copy the relevant configuration from their site configuration back into the management console. I think this is acceptable but I will ponder it some more.

felixfbecker commented on shared/gulpfile.js:
```diff
@@ -78,7 +78,10 @@ async function schema() {
     ['json-schema-draft-07', 'settings', 'critical', 'site'].map(async file => {
-      let schema = await readFile(path.join(schemaDir, `${file}.schema.json`), 'utf8')
+      let schema = await readFile(
+        path.join(schemaDir, `${file === 'critical' ? 'critical/critical' : file}.schema.json`),
```
This looks like a hack to me - why is it needed?

slimsag replied:
This is definitely a hack, I will clean this up in a follow-up PR.

slimsag replied:
Fixed in this PR via c0c48c9967f4442c64c585d9d89eb49b3beb9475

slimsag commented on .vscode/settings.json:
```diff
@@ -19,7 +18,7 @@
     {
       "fileMatch": ["dev/critical-config.json"],
-      "url": "/schema/critical.schema.json",
+      "url": "/schema/critical/critical.schema.json",
```
I will address this in a follow-up PR. The reason we need a sub-folder is kind of stupid: the Go JSON schema generator we use generates duplicate type signatures because the definitions in the schema have also been added to the `site.schema.json` which makes the Go file invalid. We can't remove them from one and share them because the schema generator also does not allow that, and renaming the type definitions in either schema would make doing the migration difficult on the Go side.

In 2 versions we can remove the critical.schema.json entirely, which will make this and many other things here a LOT cleaner.

tsenart replied:
You can define the name of the generated Go type explicitly with the title attribute. Would that help here?

slimsag replied:
As I understand, no. For example `site.schema.json` has a definition `Foobar` and `critical.schema.json` also has a definition `Foobar`, and I need them to refer to the same Go type `Foobar` (because they are the same type, and I need to easily copy one to the other in Go)

tsenart replied:
You could name them differently and then create a [type alias](https://yourbasic.org/golang/type-alias/). Would that work perhaps?

slimsag replied:
Fixed in c0c48c9967f4442c64c585d9d89eb49b3beb9475 by moving the web codebase to using the site config instead of the (now deprecated) critical config.

beyang commented:
@slimsag I've pushed my commits updating the regression tests to this branch

christinaforney commented:
This looks great! Approving from product side (though I would like to test it out). 

Is there any risk pushing this in too close to branch cut and creating problems with the release?

slimsag commented:
I am going to merge this. I think this change will be a MASSIVE improvement for our users and is unlikely to cause us any issues in the release. We have multiple escape options if I turn out to be wrong (reversion, patch releases, the documented `SITE_CONFIG_FILE` backup escape hatch, etc.)

(I've addressed all feedback (including feedback from the web team about the critical config requiring a gulpfile hack), and am committed to following up on any additional feedback anyone leaves here or in issues filed on me. There is more cleanup I can and will do to this part of our codebase overall, but this PR is too large and making those changes in it would be risky.)

There are always risks with landing large changes like this, and even more-so with this change in particular given it has not had a significant amount of time to bake (i.e. it has gone from conception -> implementation in just three days!) but I think the value we gain and risks of _not_ merging this in 3.11 greatly outweigh the risks of merging it.

Potential risks of merging this include:

- Migration trouble -> The migration is automatic so there should be no issues, and there are clear migration notes called out in the CHANGELOG, documentation in several locations, and k8s migration documentation. The risk here is minimal.
- Product change oversight (we learn we don't like the new behavior for some reason) -> this seems incredibly unlikely, as I have heard very positive feedback from multiple people on the team.
- Something falling through the cracks (this is a big change) -> It is certainly possible, and this is likely the greatest risk with this change, but I feel confident enough to merge this now. Release captain (Beyang) will also assess independently whether or not this change makes the 3.11 branch cut (along with some other things) which makes this a "two people must turn the nuclear launch key" kind of action.

The (quite large) values we gain are outlined in the PR description already, and there are large risks of this not landing in 3.11 in my view from what I have seen in our current onboarding flow and given our renewed targets for the new year.

slimsag commented:
Important note: the e2e test failure is unrelated and is passing locally, please see discussion here: https://sourcegraph.slack.com/archives/C07KZF47K/p1576555658027900?thread_ts=1576387614.018200&cid=C07KZF47K
